### ISSUE TITLE
PEP 621-style license field shows the entire license text

### ISSUE THREAD

**Issue by GergelyKalmar** [pypi/warehouse#12392]
**Describe the bug**
When using PEP 621-style project metadata with a license file (see https://peps.python.org/pep-0621/#license), the entire license text is shown in the Meta -> License section of a package on pypi. For an example see https://pypi.org/project/pytest-logikal/ and the relevant pyproject.toml file https://github.com/logikal-io/pytest-logikal/blob/main/pyproject.toml#L10. Showing the entire license text makes the Meta section almost useless by pushing the rest of the metadata quite far to the bottom.

**Expected behavior**
The license text should be hidden by default and perhaps shown with a modal upon clicking "see full text". Note that the issue is similar to https://github.com/pypi/warehouse/issues/1354, however, it is a little different due to the PEP 621 standard causing this behavior. With more and more projects adopting PEP 621 I would expect this to be a bigger issue over time.

**Comment by di** [comment]
Some related discussion is here: https://github.com/FFY00/meson-python/issues/129

**Comment by GergelyKalmar** [comment]
Thanks for the context! I think that the current UX needs to be improved regardless of how PEP 639 goes, the "Meta" section is really messed up for anyone who follows PEP 621 right now (which is the accepted standard at the moment). I think showing the (short) license name from the trove classifier there is more than enough (when available) with a fallback to the first line of the license text or so. PEP 639 would use different fields anyways, so even if that gets accepted you need to be able to deal with all the "legacy" approaches like PEP 639 or anything from before.

It seems to me that that area in the PyPI UI is not meant for displaying anything more than a short-ish line anyhow, so at the very least the excess text should be truncated/hidden.

**Comment by GergelyKalmar** [comment]
I must add that currently setuptools just copies the contents of the referenced license file into the "License" field in PKG-INFO, so it can be interpreted that the contents of the field are the actual license text of the package. I can imagine that displaying the entire license on PyPI would be potentially useful, just not in the way it is done at the moment.

**Comment by GergelyKalmar** [comment]
Lastly, I see what you mean by https://packaging.python.org/en/latest/specifications/core-metadata/#license and so I believe setuptools does the right thing, we should probably just not use the license field together with the trove classifiers (in light of the core metadata specification). Nonetheless, even if someone used the license field according to the PyPA specification and added a longer custom license text there, the Meta section on PyPI would be still messed up, so I think this is indeed a valid bug, regardless of the slightly confusing specifications surrounding license information.

**Comment by rgommers** [comment]
> Some related discussion is here: https://github.com/FFY00/meson-python/issues/129

The repo moved, here is the new link: https://github.com/mesonbuild/meson-python/issues/129

**Comment by di** [comment]
What would you like PyPI to do here that isn't already covered by #1354? We unfortunately have no way to distinguish that contents of the `License` field are due to PEP 621, just that it's a very long string that breaks our UI.

**Comment by GergelyKalmar** [comment]
It kinda works for me, even though the current implementation with just the truncation is a bit rough. I guess having a preference to just show the license as parsed from the trove classifier when there's one would be more elegant. Is there anything against doing just that?

**Comment by di** [comment]
Yes. Here's what I wrote in https://github.com/mesonbuild/meson-python/issues/129#issuecomment-1220973833:

> Because, historically, when someone specifies the `License` field, they wanted to see that value displayed on PyPI. Per the [core metadata spec](https://packaging.python.org/en/latest/specifications/core-metadata/#license), `License` is supposed to be used _instead_ of classifiers:
>
>> Text indicating the license covering the distribution where the license is not a selection from the “License” Trove classifiers.
>
> But in practice, people tend to specify both. Generally, providing multiple lines for `License` is a configuration mistake, so instead of dropping an entire file in the sidebar, PyPI just displays the first line, which is usually the name of the license (but isn't guaranteed to be).

So if you don't want to see the truncated bit, the solution is to not specify it as `License` and just use the classifiers instead.

**Comment by GergelyKalmar** [comment]
I understand, it's just that this approach seems to be very counter-intuitive in light of PEP 621. What you are saying essentially means we _must not_ use the `license` field as recommended there whenever we use a standard license. That is really confusing.

Perhaps what was proposed in #1354 would indeed be the most elegant solution after all. Having a link saying "see full license text" instead of the current truncated sentence seems like the most user- and package author-friendly approach.

**Comment by rgommers** [comment]
I noticed that this is now working much better. I don't know if it was fixed on purpose, and if so, by whom and how. But thank you either way:)

This issue can probably be closed now? Here is what it now looks like for SciPy (`pytest-logikal` and other packages I checked look good too):


<img width="869" alt="image" src="https://github.com/pypi/warehouse/assets/98330/809ab547-392b-42f6-ac48-225f07e71fa5">
